File: 10-algo-ds-implementations/union_find_rank_heuristic.py

class Union_Find_Rank_Heuristic:
    def __init__(self, n: int):
        self.__rank = [ 0 for _ in range(n) ]
        self.__parent = [ i for i in range(n) ]
    
    # Time Complexity: O(logn)
    def find(self, i: int) -> int:
        if i < 0 or i >= len(self.__parent):
            return -1
        
        root_i = i
        while root_i != self.__parent[root_i]:
            root_i = self.__parent[root_i]
        
        return root_i
    
    # Time Complexity: O(logn)
    def union(self, i: int, j: int):
        root_i = self.find(i)
        root_j = self.find(j)
        if root_i == root_j:
            return
        
        if self.__rank[root_i] > self.__rank[root_j]:
            # We choose root_i as the root of the merged tree
            self.__parent[root_j] = root_i
         
        else:
            # We choose root_j as the root of the merged tree
            self.__parent[root_i] = root_j
            if self.__rank[root_i] == self.__rank[root_j]:
                self.__rank[root_j] += 1
        

File: 09-typescript/algo-ds-implementations/unionFindRankHeuristic.ts


export class UnionFindRankHeuristic {
    private _rank: number[]
    private _parent: number[]

    public constructor(size: number) {
        this._rank = []
        this._parent = []

        for(let i = 0; i < size; i++) {
            this._rank.push(0)
            this._parent.push(i)
        }
    }

    public find(i: number): number {
        if (i < 0 || i >= this._parent.length) {
            return -1
        }

        let iRoot = i
        while (iRoot !== this._parent[iRoot]) {
            iRoot = this._parent[iRoot]
        }

        return iRoot
    }

    public union(i: number, j: number): void {
        let iRoot = this.find(i)
        let jRoot = this.find(j)
        if (iRoot === jRoot) {
            return
        }

        if (this._rank[iRoot] < this._rank[jRoot]) {
            this._parent[iRoot] = jRoot

        } else {
            this._parent[jRoot] = iRoot

            if (this._rank[iRoot] === this._rank[jRoot]) {
                this._rank[iRoot] += 1
            }
        }
    }
}